### PR TITLE
Add SplineC2C OpenMP offload

### PR DESCRIPTION
**Pull request by ye-luo**
## Proposed changes
Implement complex spline orbital code path. Existing unit and integration tests cover the added code paths.

## What type(s) of changes does this code introduce?
- New feature

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
bora-external

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted

### CODE REVIEW

File: src/QMCWaveFunctions/BsplineFactory/SplineC2COMP.cpp [hunk 1]
```diff
@@ -0,0 +1,1294 @@
...
+  const ST gY_i = dY_i - val_r * kY;
+  const ST gZ_i = dZ_i - val_r * kZ;
+
+  const ST lcart_r = SymTrace(h00[jr], h01[jr], h02[jr], h11[jr], h12[jr], h22[jr], symGGt);
+  const ST lcart_i = SymTrace(h00[ji], h01[ji], h02[ji], h11[ji], h12[ji], h22[ji], symGGt);
+  const ST lap_r   = lcart_r + mKK_ptr[index] * val_r + two * (kX * dX_i + kY * dY_i + kZ * dZ_i);
+  const ST lap_i   = lcart_i + mKK_ptr[index] * val_i - two * (kX * dX_r + kY * dY_r + kZ * dZ_r);
+
+  const size_t psiIndex = first_spo + index;
+  //this will be fixed later
```
Comment (prckent) [inline src/QMCWaveFunctions/BsplineFactory/SplineC2COMP.cpp:128]: What will be fixed?

File: src/QMCWaveFunctions/BsplineFactory/SplineC2COMP.cpp [hunk 2]
```diff
@@ -0,0 +1,1294 @@
...
+  psi[psiIndex]   = TT(c * val_r - s * val_i, c * val_i + s * val_r);
+  d2psi[psiIndex] = TT(c * lap_r - s * lap_i, c * lap_i + s * lap_r);
+  //this will go way with Determinant
```
Comment (prckent) [inline src/QMCWaveFunctions/BsplineFactory/SplineC2COMP.cpp:131]: What will go away?

File: src/QMCWaveFunctions/BsplineFactory/SplineC2COMP.cpp [hunk 3]
```diff
@@ -0,0 +1,1294 @@
...
+        const int last_cplx  = orb_size < last / 2 ? orb_size : last / 2;
+        ComplexT sum(0);
+        PRAGMA_OFFLOAD("omp parallel for simd reduction(+:sum)")
+        for (int i = first_cplx; i < last_cplx; i++)
+          sum += psi_iat_ptr[i] * psiinv_ptr[i];
+        ratios_private_ptr[iat * NumTeams + team_id] = sum;
+      }
+  }
+
+  // do the reduction manually
```
Comment (prckent) [inline src/QMCWaveFunctions/BsplineFactory/SplineC2COMP.cpp:296]: Why? (I think we might be evading a compiler weakness. Under what conditions do we revisit?)

File: src/QMCWaveFunctions/BsplineFactory/createComplexDouble.cpp [hunk 4]
```diff
@@ -36,10 +37,24 @@ BsplineReaderBase* createBsplineComplexDouble(EinsplineSetBuilder* e, bool hybri
...
 
 #if defined(QMC_COMPLEX)
-  if (hybrid_rep)
-    aReader = new HybridRepSetReader<HybridRepCplx<SplineC2C<double>>>(e);
+#if defined(ENABLE_OFFLOAD)
+  if (useGPU == "yes")
+  {
+    if (hybrid_rep)
+    {
+      APP_ABORT("OpenMP offload has not been enabled with hybrid orbital representation!");
```
Comment (prckent) [inline src/QMCWaveFunctions/BsplineFactory/createComplexDouble.cpp:45]: The new code must always work and never abort if offload code is not present. The CPU path should be used when hybrid rep is selected for now. Same goes for real splines. Create an issue if you don't want to fix this now.

File: src/QMCWaveFunctions/BsplineFactory/createComplexSingle.cpp [hunk 5]
```diff
@@ -36,10 +37,24 @@ BsplineReaderBase* createBsplineComplexSingle(EinsplineSetBuilder* e, bool hybri
...
 
 #if defined(QMC_COMPLEX)
-  if (hybrid_rep)
-    aReader = new HybridRepSetReader<HybridRepCplx<SplineC2C<float>>>(e);
+#if defined(ENABLE_OFFLOAD)
+  if (useGPU == "yes")
+  {
+    if (hybrid_rep)
+    {
+      APP_ABORT("OpenMP offload has not been enabled with hybrid orbital representation!");
```
Comment (prckent) [inline src/QMCWaveFunctions/BsplineFactory/createComplexSingle.cpp:45]: Same comment as the non-single implementation. The code should use the CPU path in this case.

File: src/QMCWaveFunctions/BsplineFactory/SplineC2COMP.cpp (hunk 1 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/BsplineFactory/SplineC2COMP.cpp]: Removed outdated comments

File: src/QMCWaveFunctions/BsplineFactory/SplineC2COMP.cpp (hunk 2 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/BsplineFactory/SplineC2COMP.cpp]: Removed outdated comments

File: src/QMCWaveFunctions/BsplineFactory/SplineC2COMP.cpp (hunk 3 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/BsplineFactory/SplineC2COMP.cpp:294]: GPU reduction across SM/CU is complicated. So done on the host.

File: src/QMCWaveFunctions/BsplineFactory/createComplexDouble.cpp (hunk 4 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/BsplineFactory/createComplexDouble.cpp]: No reason to need more user actions. Fallback to host code path.

File: src/QMCWaveFunctions/BsplineFactory/createComplexSingle.cpp (hunk 5 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/BsplineFactory/createComplexSingle.cpp]: No reason to need more user actions. Fallback to host code path.

**Comment by prckent** [comment]
Thanks for the updates. Q. What happens with hybrid rep, real build, and offload?

**Comment by ye-luo** [comment]
> 
> 
> Thanks for the updates. Q. What happens with hybrid rep, real build, and offload?

complex and real are treated the same. When hybrid and offload are both requested. The CPU code path is used with a message indicating the reason.

File: src/QMCWaveFunctions/BsplineFactory/createComplexSingle.cpp [hunk 6]
```diff
@@ -36,17 +37,35 @@ BsplineReaderBase* createBsplineComplexSingle(EinsplineSetBuilder* e, bool hybri
...
 #else //QMC_COMPLEX
 #if defined(ENABLE_OFFLOAD)
   if (useGPU == "yes")
   {
     if (hybrid_rep)
     {
-      APP_ABORT("OpenMP offload has not been enabled with hybrid orbital representation!");
+      app_log() << "OpenMP offload has not been enabled with hybrid orbital representation!"
+                << " Running on the host." << std::endl;
+      aReader = new HybridRepSetReader<HybridRepCplx<SplineC2R<float>>>(e);
```
Comment (ye-luo) [inline src/QMCWaveFunctions/BsplineFactory/createComplexSingle.cpp:68]: @prckent this is the real case and CPU fallback is connected.

**Pull request merged**